### PR TITLE
Add span links support to OpenTelemetry instrumentation

### PR DESCRIPTION
**Pull request by PerfectSlayer**
# What Does This Do

This PR adds OpenTelemetry span links support for its instrumentation.

# Motivation

This should improve OpenTelemetry API support and compatibility.

# Additional Notes

This relies on [the past span links to core addition](https://github.com/DataDog/dd-trace-java/pull/5569) and [its update](https://github.com/DataDog/dd-trace-java/pull/6009).

JIRA ticket: [APMJAVA-1087]

[APMJAVA-1087]: https://datadoghq.atlassian.net/browse/APMJAVA-1087?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by PerfectSlayer** [comment]
This PR was pushed back as draft as the underlying API / encoding has to change.

File: dd-java-agent/instrumentation/opentelemetry/opentelemetry-1.4/src/test/groovy/OpenTelemetry14Test.groovy [hunk 1]
```diff
@@ -151,6 +154,154 @@ class OpenTelemetry14Test extends AgentTestRunner {
...
+          }
+        }
+      }
+    }
+
+    where:
+    attributes | expectedAttributes
+    Attributes.empty() | null
+    Attributes.builder().put("string-key", "string-value").put("long-key", 123456789L).put("double-key", 1234.5678D).put("boolean-key-true", true).put("boolean-key-false", false).build() | '{ string-key: "string-value", long-key: "123456789", double-key: "1234.5678", boolean-key-true: "true", boolean-key-false: "false" }'
+    Attributes.builder().put("string-key-array", "string-value1", "string-value2", "string-value3").put("long-key-array", 123456L, 1234567L, 12345678L).put("double-key-array", 1234.5D, 1234.56D, 1234.567D).put("boolean-key-array", true, false, true).build() | '{ string-key-array.0: "string-value1", string-key-array.1: "string-value2", string-key-array.2: "string-value3", long-key-array.0: "123456", long-key-array.1: "1234567", long-key-array.2: "12345678", double-key-array.0: "1234.5", double-key-array.1: "1234.56", double-key-array.2: "1234.567", boolean-key-array.0: "true", boolean-key-array.1: "false", boolean-key-array.2: "true" }'
```
Comment (bantonsson) [inline dd-java-agent/instrumentation/opentelemetry/opentelemetry-1.4/src/test/groovy/OpenTelemetry14Test.groovy:262]: The encoding here confuses me. It looks like a mix of JSON and the exploded tag array encoding. I thought that we should use one or the other?

Comment (PerfectSlayer) [reply dd-java-agent/instrumentation/opentelemetry/opentelemetry-1.4/src/test/groovy/OpenTelemetry14Test.groovy:262]: According the RFC, we do use both:
* We encode the span link sequenced collection using a JSON array,
* Each element is a JSON object that described the span link,
* Inside a span link element, there could be a JSON field name `attributes` that contains a JSON object with all the span link attributes,
*  Such span links attributes are sliced when being of an array type.

So you end up with both JSON (the overall encoding) and exploded array encoding.
The slicing is required independently of the encoding (v0.4 or JSON).

Comment (bantonsson) [reply dd-java-agent/instrumentation/opentelemetry/opentelemetry-1.4/src/test/groovy/OpenTelemetry14Test.groovy:262]: I can't see that being done for [OTLP in the Datadog Agent](https://github.com/DataDog/datadog-agent/blob/bfdff76022461a1f4388410f9e75ad034e787962/pkg/trace/api/otlp.go#L418)

Comment (PerfectSlayer) [reply dd-java-agent/instrumentation/opentelemetry/opentelemetry-1.4/src/test/groovy/OpenTelemetry14Test.groovy:267]: Discussed here: https://github.com/DataDog/dd-trace-java/pull/6009#discussion_r1380282662

**Review by mcculls — APPROVED** [review]

**Pull request merged**